### PR TITLE
nginx: support basic auth in location blocks

### PR DESCRIPTION
**Pull request by grahamc**
###### Motivation for this change

Extend the nginx support for basic auth to the location blocks.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by aanderse — APPROVED** [review]
Usually I'm a pretty big grump for passwords in the nix store, but if you're doing basic auth you probably aren't using it for real protection anyways... and I'm certainly guilty of having passwords in `httpd` configs under `/etc` that are world readable (on Debian, at least).

I guess I should whip something together for the `httpd` module so I can stay keeping up with the Joneses...

File: nixos/modules/services/web-servers/nginx/location-options.nix [hunk 1]
```diff
@@ -9,6 +9,31 @@ with lib;
...
+        nix store.
+      '';
+    };
+
+    basicAuthFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        Basic Auth password file for a vhost.
+        Can be created via: <command>htpasswd -c &lt;filename&gt; &lt;username&gt;</command>
```
Comment (ajs124) [inline nixos/modules/services/web-servers/nginx/location-options.nix:33]: Maybe it should be mentioned that the hashes htpasswd produces by default and all the hashes nginx actually understands aren't really considered good for storing passwords anymore.

File: nixos/tests/all-tests.nix [hunk 2]
```diff
@@ -243,6 +243,7 @@ in
   nfs4 = handleTest ./nfs { version = 4; };
   nghttpx = handleTest ./nghttpx.nix {};
   nginx = handleTest ./nginx.nix {};
+  nginx-auth= handleTest ./nginx-auth.nix {};
```
Comment (ajs124) [inline nixos/tests/all-tests.nix:246]: missing space

File: nixos/modules/services/web-servers/nginx/location-options.nix (hunk 1 above)
Comment (grahamc) [reply nixos/modules/services/web-servers/nginx/location-options.nix:33]: Possibly so, what would you like the wording to be? Note: these options and their text mirror the vhost-options.nix content.

Comment (ajs124) [reply nixos/modules/services/web-servers/nginx/location-options.nix:33]: Right. I mean, in an ideal world I would like this to say "use `htpasswd -B` to create bcrypt hashed passwords", but nginx doesn't support those.

So maybe just something along the lines of: "Warning: The generate file contains the users' passwords in a non-cryptographically-securely hashed way"

File: nixos/modules/services/web-servers/nginx/location-options.nix [hunk 3]
```diff
@@ -9,6 +9,31 @@ with lib;
...
+      '';
```
Comment (Mic92) [inline nixos/modules/services/web-servers/nginx/location-options.nix:34]:
```suggestion
      WARNING: The generate file contains the users' passwords in a non-cryptographically-securely hashed way
      '';
```

**Review by Mic92 — APPROVED** [review]

**Comment by grahamc** [comment]
I added that warning and squashed the fixup. Thanks! I'll merge once the PR is green.

**Pull request merged**